### PR TITLE
FINERACT-2065: Schedule handling for fixed length configuration

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

Following with the Loan Product `Fixed Length` configuration we need to consider the schedule generation with this new setting.

We will get this with the next considerations:
- Should be supported for zero interest bearing loan accounts only
- Should be available for “advance payment allocation“
- Validation should be in place when the fixed length cannot derive the no of repayments entered - system should not approve such config and we are creating a new `BadFixedLengthException` 

[FINERACT-2065](https://github.com/apache/fineract/pull/1284).

https://github.com/apache/fineract/assets/154766431/86433e22-628b-4014-83f5-061ad73ac220

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Review by reluxa — COMMENTED** [review]
Could you please add some unit / integration tests?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java [hunk 1]
```diff
@@ -58,6 +61,7 @@ public LoanScheduleModel generate(final MathContext mc, final LoanApplicationTer
         final ApplicationCurrency applicationCurrency = loanApplicationTerms.getApplicationCurrency();
         // generate list of proposed schedule due dates
         LocalDate loanEndDate = getScheduledDateGenerator().getLastRepaymentDate(loanApplicationTerms, holidayDetailDTO);
+        log.info("loanEndDate {}", loanEndDate);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Please remove this info logger...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java [hunk 2]
```diff
@@ -102,6 +106,19 @@ public LoanScheduleModel generate(final MathContext mc, final LoanApplicationTer
             // Loan Schedule Exceptions that need to be applied for Loan Account
             LoanTermVariationParams termVariationParams = applyLoanTermVariations(loanApplicationTerms, scheduleParams, scheduledDueDate);
             scheduledDueDate = termVariationParams.scheduledDueDate();
+            log.info("scheduledDueDate {}", scheduledDueDate);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Please remove this info logger...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java [hunk 3]
```diff
@@ -102,6 +106,19 @@ public LoanScheduleModel generate(final MathContext mc, final LoanApplicationTer
...
+            if (loanApplicationTerms.getFixedLength() != null) {
+                if (loanApplicationTerms.getNumberOfRepayments().compareTo(scheduleParams.getInstalmentNumber()) == 0) {
+                    if (DateUtils.isBefore(scheduledDueDate, loanEndDate)) {
+                        scheduledDueDate = loanEndDate;
+                    }
+                }
+
+                isNextRepaymentAvailable = DateUtils.isBefore(scheduledDueDate, loanEndDate);
+            }
+            log.info("scheduledDueDate {}", scheduledDueDate);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Please remove this info logger...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java [hunk 4]
```diff
@@ -128,6 +145,16 @@ public LoanScheduleModel generate(final MathContext mc, final LoanApplicationTer
 
             if (!isNextRepaymentAvailable) {
                 scheduleParams.getDisburseDetailMap().clear();
+
+                // Fixed Length Validations - If the Number or Repayments requested is greater than current/last
+                // Installment number
+                if (loanApplicationTerms.getFixedLength() != null
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: This should be validated in the validator, not during the schedule generation.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java [hunk 5]
```diff
@@ -102,6 +106,19 @@ public LoanScheduleModel generate(final MathContext mc, final LoanApplicationTer
             // Loan Schedule Exceptions that need to be applied for Loan Account
             LoanTermVariationParams termVariationParams = applyLoanTermVariations(loanApplicationTerms, scheduleParams, scheduledDueDate);
             scheduledDueDate = termVariationParams.scheduledDueDate();
+            log.info("scheduledDueDate {}", scheduledDueDate);
+
+            // Fixed Length Validations - Ensure the last repayment date equal to Fixed Length
+            if (loanApplicationTerms.getFixedLength() != null) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: I dont fully understand this here...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 6]
```diff
@@ -46,17 +45,26 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
...
             isFirstRepayment = false;
+            // Fixed Length validation - When the Number of Repayments is greater than Fixed Length
+            if (maxDateForFixedLength != null && DateUtils.isAfter(lastRepaymentDate, maxDateForFixedLength)) {
+                lastRepaymentDate = maxDateForFixedLength;
+                break;
+            }
         }
         lastRepaymentDate = adjustRepaymentDate(lastRepaymentDate, loanApplicationTerms, holidayDetailDTO).getChangedScheduleDate();
+        // Fixed Length validation - When the Number of Repayments is lower than Fixed Length
+        if (maxDateForFixedLength != null && DateUtils.isAfter(maxDateForFixedLength, lastRepaymentDate)) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Why do we do it twice?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 7]
```diff
@@ -46,17 +45,26 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
...
+            lastRepaymentDate = maxDateForFixedLength;
+        }
+        log.info("lastRepaymentDate {}", lastRepaymentDate);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Please remove the info logger

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/exception/BadFixedLengthSetupException.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanaccount.loanschedule.exception;
+
+import org.apache.fineract.infrastructure.core.exception.AbstractPlatformDomainRuleException;
+
+public class BadFixedLengthSetupException extends AbstractPlatformDomainRuleException {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/exception/BadFixedLengthSetupException.java]: We dont need a new exception. Please use the GenericPlatformDomainRuleException

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 9]
```diff
@@ -3467,6 +3467,147 @@ public void uc124() {
...
+                    new PostLoansLoanIdRequest().actualDisbursementDate("22 November 2023").dateFormat(DATETIME_PATTERN)
+                            .transactionAmount(BigDecimal.valueOf(100.0)).locale("en"));
+
+            GetLoansLoanIdResponse loanDetails = loanTransactionHelper.getLoanDetails(loanResponse.getLoanId());
+            validateLoanSummaryBalances(loanDetails, 100.0, 0.0, 100.0, 0.0, null);
+            validateRepaymentPeriod(loanDetails, 1, LocalDate.of(2023, 12, 7), 33.0, 0.0, 33.0, 0.0, 0.0);
+            validateRepaymentPeriod(loanDetails, 2, LocalDate.of(2023, 12, 22), 33.0, 0.0, 33.0, 0.0, 0.0);
+            validateRepaymentPeriod(loanDetails, 3, LocalDate.of(2024, 1, 01), 34.0, 0.0, 34.0, 0.0, 0.0);
+            assertEquals(loanDetails.getNumberOfRepayments(), 3);
+            assertEquals(Utils.getDifferenceInDays(loanDetails.getTimeline().getActualDisbursementDate(), LocalDate.of(2024, 1, 01)),
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: You should test the due date of the very last installment instead of hard coding a date...

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 10]
```diff
@@ -3467,6 +3467,147 @@ public void uc124() {
...
+                    new PostLoansLoanIdRequest().actualDisbursementDate("22 November 2023").dateFormat(DATETIME_PATTERN)
+                            .transactionAmount(BigDecimal.valueOf(100.0)).locale("en"));
+
+            GetLoansLoanIdResponse loanDetails = loanTransactionHelper.getLoanDetails(loanResponse.getLoanId());
+            validateLoanSummaryBalances(loanDetails, 100.0, 0.0, 100.0, 0.0, null);
+            validateRepaymentPeriod(loanDetails, 1, LocalDate.of(2023, 12, 6), 33.0, 0.0, 33.0, 0.0, 0.0);
+            validateRepaymentPeriod(loanDetails, 2, LocalDate.of(2023, 12, 20), 33.0, 0.0, 33.0, 0.0, 0.0);
+            validateRepaymentPeriod(loanDetails, 3, LocalDate.of(2023, 12, 27), 34.0, 0.0, 34.0, 0.0, 0.0);
+            assertEquals(loanDetails.getNumberOfRepayments(), 3);
+            assertEquals(Utils.getDifferenceInWeeks(loanDetails.getTimeline().getActualDisbursementDate(), LocalDate.of(2023, 12, 27)),
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: Same as above

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 11]
```diff
@@ -3467,6 +3467,147 @@ public void uc124() {
...
+                    fixedLength.longValue());
+            assertTrue(loanDetails.getStatus().getActive());
+        });
+    }
+
+    // UC127: Advanced payment allocation with Fixed Length for 11 months and Loan Term for 12 months (6 repayments
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: Please add a test case when the fixed lenght is longer than the loan term.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 12]
```diff
@@ -3467,6 +3467,147 @@ public void uc124() {
...
+            PostLoansResponse loanResponse = loanTransactionHelper.applyLoan(applicationRequest);
+
+            loanTransactionHelper.approveLoan(loanResponse.getLoanId(),
+                    new PostLoansLoanIdRequest().approvedLoanAmount(BigDecimal.valueOf(1000)).dateFormat(DATETIME_PATTERN)
+                            .approvedOnDate("22 November 2023").locale("en"));
+
+            loanTransactionHelper.disburseLoan(loanResponse.getLoanId(),
+                    new PostLoansLoanIdRequest().actualDisbursementDate("22 November 2023").dateFormat(DATETIME_PATTERN)
+                            .transactionAmount(BigDecimal.valueOf(100.0)).locale("en"));
+
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java:3675]: Please add test cases where you are testing negative scenarios:
- Providing invalid fixed length parameters

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my comments!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Done, Removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Done, Removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Done, Removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: We have talked about this, that validation will be keep it here

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 5 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: We have talked about this in the session

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 6 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: We have talked about this in the session

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 7 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Done, Removed!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/exception/BadFixedLengthSetupException.java (hunk 8 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/exception/BadFixedLengthSetupException.java]: Done! removed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 13]
```diff
@@ -46,15 +45,22 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
...
+        final LocalDate maxDateForFixedLength = loanApplicationTerms.calculateMaxDateForFixedLength();
 
-        LocalDate lastRepaymentDate = RepaymentStartDateType.DISBURSEMENT_DATE.equals(repaymentStartDateType)
-                ? loanApplicationTerms.getExpectedDisbursementDate()
-                : loanApplicationTerms.getSubmittedOnDate();
         boolean isFirstRepayment = true;
         for (int repaymentPeriod = 1; repaymentPeriod <= numberOfRepayments; repaymentPeriod++) {
             lastRepaymentDate = generateNextRepaymentDate(lastRepaymentDate, loanApplicationTerms, isFirstRepayment);
             isFirstRepayment = false;
+            // Fixed Length validation - When the Number of Repayments is greater than Fixed Length
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Please correct these comments. This has nothing to do with the number of repayments... it checks whether the last repaymentDate is greater than the fixedLengthDate...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 14]
```diff
@@ -46,15 +45,22 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
...
+            if (maxDateForFixedLength != null && DateUtils.isAfter(lastRepaymentDate, maxDateForFixedLength)) {
+                lastRepaymentDate = maxDateForFixedLength;
+                break;
+            }
+        }
+        // Fixed Length validation - When the Number of Repayments is lower than Fixed Length
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Please correct these comments. This has nothing to do with the number of repayments... it checks whether the last repaymentDate is lower than the fixedLengthDate...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 15]
```diff
@@ -46,15 +45,22 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
...
-                ? loanApplicationTerms.getExpectedDisbursementDate()
-                : loanApplicationTerms.getSubmittedOnDate();
         boolean isFirstRepayment = true;
         for (int repaymentPeriod = 1; repaymentPeriod <= numberOfRepayments; repaymentPeriod++) {
             lastRepaymentDate = generateNextRepaymentDate(lastRepaymentDate, loanApplicationTerms, isFirstRepayment);
             isFirstRepayment = false;
+            // Fixed Length validation - When the Number of Repayments is greater than Fixed Length
+            if (maxDateForFixedLength != null && DateUtils.isAfter(lastRepaymentDate, maxDateForFixedLength)) {
+                lastRepaymentDate = maxDateForFixedLength;
+                break;
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Would you mind moving this condition out of the cycle, it does not require do to anything there. After the cycle finished you will have the `lastRepaymentDate` set, which is the due date of the last installment and after you can decide whether it needs to be changed or not:
if `lastRepaymentDate` is not equal with the `fixedLengthDate`, override it with the `fixedLengthDate`.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 6 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: See https://github.com/apache/fineract/pull/3808/files#r1532214898

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 13 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 14 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 15 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Done!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java (hunk 11 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: TC uc128 created

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java (hunk 10 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: Done, validations updated with the due date of the last period

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java (hunk 9 above)
Comment (alberto-art3ch) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: Done, validations updated with the due date of the last period

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 4 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Removed from this method and class

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/AbstractProgressiveLoanScheduleGenerator.java]: Removed from this method and class

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 16]
```diff
@@ -104,6 +104,12 @@ public LocalDate generateNextRepaymentDate(final LocalDate lastRepaymentDate, fi
             }
         }
 
+        final LocalDate maxDateForFixedLength = loanApplicationTerms.calculateMaxDateForFixedLength();
+        // Fixed Length validation
+        if (maxDateForFixedLength != null && DateUtils.isAfter(dueRepaymentPeriodDate, maxDateForFixedLength)) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:108]: I dont think this is righ. I reckon we might have validation to avoid such situation where we are adjusting not the last repayment, but still this is not checking whether it is the latest or not. 

Also i dont think we need this here... Its more than enough to adjust the loan end date and the very last installment due date, no?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 17]
```diff
@@ -46,15 +45,16 @@ public LocalDate getLastRepaymentDate(final LoanApplicationTerms loanApplication
...
-
-        LocalDate lastRepaymentDate = RepaymentStartDateType.DISBURSEMENT_DATE.equals(repaymentStartDateType)
-                ? loanApplicationTerms.getExpectedDisbursementDate()
-                : loanApplicationTerms.getSubmittedOnDate();
-        boolean isFirstRepayment = true;
-        for (int repaymentPeriod = 1; repaymentPeriod <= numberOfRepayments; repaymentPeriod++) {
-            lastRepaymentDate = generateNextRepaymentDate(lastRepaymentDate, loanApplicationTerms, isFirstRepayment);
-            isFirstRepayment = false;
+        LocalDate lastRepaymentDate = loanApplicationTerms.getRepaymentStartDate();
+        // When exists defined a Fixed Length value
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: We dont really need this..

Please revert and just adjust the lastRepaymentDate after it got calculated (the after the for cycle) in the following manner:
if the `lastRepaymentDate` is not equals with `calculatedMaxDateForFixedLenght` than override the `lastRepaymentDate` with it.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Kindly check my review and please add the requested negative test scenarios!

**Comment by alberto-art3ch** [comment]
> Kindly check my review and please add the requested negative test scenarios!

The last change was already done and the negative case is the uc128

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 16 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:108]: Done! updated

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 17 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java]: Done! updated

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java [hunk 18]
```diff
@@ -3467,6 +3467,181 @@ public void uc124() {
...
+    public void uc128() {
+        final String operationDate = "22 November 2023";
+        runAt(operationDate, () -> {
+            final Integer repaymentFrequencyType = 2; // month
+
+            PostLoanProductsRequest product = createOnePeriod30DaysLongNoInterestPeriodicAccrualProductWithAdvancedPaymentAllocation()
+                    .numberOfRepayments(6).repaymentEvery(1).repaymentFrequencyType(repaymentFrequencyType.longValue()).fixedLength(6);
+            PostLoanProductsResponse loanProductResponse = loanProductHelper.createLoanProduct(product);
+
+            loanTransactionHelper.applyLoanWithError(new PostLoansRequest().clientId(client.getClientId())
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/AdvancedPaymentAllocationLoanRepaymentScheduleTest.java]: You should check whether the provided exception and message is correct or not.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java (hunk 16 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:108]: I dont see any changes here... why do we need this for the "generateNextRepaymentDate"?

I dont think we need this at all. Please let me know if i am mistaken.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments and kindly fix the conflicts!

**Comment by adamsaghy** [comment]
@alberto-art3ch Please fix the code format error!

**Review by adamsaghy — APPROVED** [review]
LGTM

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java [hunk 19]
```diff
@@ -2394,4 +2405,44 @@ private void validatePartialPeriodSupport(final Integer interestCalculationPerio
...
+                            errorMsg, AdvancedPaymentScheduleTransactionProcessor.ADVANCED_PAYMENT_ALLOCATION_STRATEGY);
+                }
+
+                if (interestRatePerPeriod.compareTo(BigDecimal.ZERO) > 0) {
+                    final String errorMsg = "Fixed Length configuration is only allowed for zero interest products";
+                    throw new GeneralPlatformDomainRuleException("error.msg.fixed.length.only.supported.for.zero.interest", errorMsg,
+                            interestRatePerPeriod);
+                }
+
+                if (fixedLength != null && numberOfRepayments.compareTo(fixedLength) > 0) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java]: This is incorrect...
Example which should work:
- number of repayments: 3
- Repay every: 15 DAYS
- Loan term: 45 DAYS
- Fixed term length: 40 days

**Review by adamsaghy — CHANGES_REQUESTED** [review]
https://github.com/apache/fineract/pull/3808/files#r1539759545

**Review by adamsaghy — APPROVED** [review]
LGTM
